### PR TITLE
Embedding columns (from embedding providers) are now run inside datafusion plans.

### PR DESCRIPTION
**Pull request by Jeadie**
## Changes
- New key in the dataset to determine which columns should be augmented with an embedding. For example
```yaml
datasets:
  - from: s3://spiceai-demo-datasets/cleaned_sales_data.parquet
    name: embeded_cleaned_sales_data
    embeddings:
      - column: address_line1
        use: oai
    params:
      file_format: parquet

embeddings:
  - name: oai
    from: openai
```

- New `impl TableProvider`, `EmbeddingTable` that wraps an arbitrary `Arc<dyn TableProvider>` and adds the specified new columns.
  - For column, `col`, embedding column is `col_embedding`.
- New `impl DataConnector`, `EmbeddingConnector` that wraps any arbitrary `Arc<dyn DataConnector>` and produces `EmbeddingTable` TableProviders.
  - i.e. If `Arc<dyn DataConnector> -> Arc<dyn TableProvider>`, then `EmbeddingConnector(Arc<dyn DataConnector>) -> EmbeddingTable(Arc<dyn TableProvider>)`
- New `EmbeddingTableExec`, `impl ExecutionPlan` to perform the scan for an `EmbeddingTable`.
- New method in `Embed` trait: `fn size(&self) -> i32;`.

## Other Changes
- Load models/embeddings/LLMs before datasets so that the embedding models can be available to precompute dataset types correctly (e.g. length of vectors).
- Bug in `ObjectStoreMetadataTable` and `ObjectStoreTextTable` regarding limits. Currently we increment our limit check for each record batch, not the number of elements per batch.

### CODE REVIEW

**Comment by Jeadie** [comment]
#### Bug 1
- Querying on an embedding column without a `LIMIT n` clause fails with error

```
sql> select address_line1_embedding  from embeded_cleaned_sales_data
Error An internal error occurred. Execute '.error' to show details.

sql> .error
Status { code: Internal, message: "Unable to collect results after query execution: OutputRequirements\ncaused by\nInternal error: PhysicalOptimizer rule 'OutputRequirements' failed, due to generate a different schema, original schema: Schema { fields: [Field { name: \"address_line1_embedding\", data_type: FixedSizeList(Field { name: \"item\", data_type: Float32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, 1536), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, new schema: Schema { fields: [Field { name: \"address_line1\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }.\nThis was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker", metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Sun, 02 Jun 2024 11:34:59 GMT", "content-length": "0"} }, source: None }

sql> select address_line1_embedding  from embeded_cleaned_sales_data limit 2

+-----------------------------------+
| address_line1_embedding           |
+-----------------------------------+
|           [1.0, 0.3, ... ]        |
|           [ 4.2, 0.1, ...]        |
+-----------------------------------+
```

**Comment by Jeadie** [comment]
#### Bug 2
- Acceleration does not work. On startup with 
```yaml
datasets:
  - from: s3://spiceai-demo-datasets/cleaned_sales_data.parquet
    name: embeded_cleaned_sales_data
    acceleration:
      enabled: true
    embeddings:
      - column: address_line1
        use: oai
    params:
      file_format: parquet

embeddings:
  - name: oai
    from: openai
```

An error is produced

```
Internal error: PhysicalOptimizer rule 'OutputRequirements' failed, due to generate a different schema, original schema: Schema { fields: [Field { name: "order_number", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "quantity_ordered", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "price_each", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "order_line_number", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "sales", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "order_date", data_type: Timestamp(Microsecond, None), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "status", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "quarter", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "month", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "year", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "product_line", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "msrp", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "product_code", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "customer_name", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "phone", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "address_line1", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "address_line2", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "city", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "state", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "postal_code", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "country", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "territory", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "contact_last_name", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "contact_first_name", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "deal_size", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "address_line1_embedding", data_type: FixedSizeList(Field { name: "item", data_type: Float32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, 1536), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, new schema: Schema { fields: [Field { name: "order_number", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "quantity_ordered", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "price_each", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "order_line_number", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "sales", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "order_date", data_type: Timestamp(Microsecond, None), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "status", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "quarter", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "month", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "year", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "product_line", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "msrp", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "product_code", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "customer_name", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "phone", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "address_line1", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "address_line2", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "city", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "state", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "postal_code", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "country", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "territory", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "contact_last_name", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "contact_first_name", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "deal_size", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

The core part of the error is the `old schema` has the `address_line1_embedding`, and the new schema does not.

```
Internal error: PhysicalOptimizer rule 'OutputRequirements' failed, due to generate a different schema, original schema:
```

File: crates/runtime/src/lib.rs [hunk 1]
```diff
@@ -411,7 +413,10 @@ impl Runtime {
             }
         };
 
-        Ok(data_connector)
+        Ok(Arc::new(EmbeddingConnector::new(
+            data_connector,
+            Arc::clone(&self.embeds),
+        )) as Arc<dyn DataConnector>)
```
Comment (phillipleblanc) [inline crates/runtime/src/lib.rs]: I would only "wrap" the connector with the EmbeddingConnector if it needs it. It's not obvious from this code that it is a no-op if there are no embeddings.

Comment (phillipleblanc) [reply crates/runtime/src/lib.rs]: Yeah I would rework this logic to handle the wrapping in the register_dataset function.

File: crates/runtime/src/embeddings/table.rs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    async fn precompute_embedding_sizes(
+        embedded_columns: &HashMap<String, String>,
+        embedding_models: &Arc<RwLock<EmbeddingModelStore>>,
+    ) -> HashMap<String, i32> {
+        let mut model_sizes: HashMap<String, i32> = HashMap::new();
+        for (col, model) in embedded_columns {
+            if let Some(model_lock) = embedding_models.read().await.get(model) {
+                let z = model_lock.read().await.size();
+                model_sizes.insert(col.clone(), z);
+                println!("Model {model} has size {z}");
```
Comment (phillipleblanc) [inline crates/runtime/src/embeddings/table.rs]:
```suggestion
                tracing::debug!("Model {model} has size {z}");
```

File: crates/runtime/src/embeddings/table.rs [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            } else {
+                println!("Model {model} not found for column {col}");
```
Comment (phillipleblanc) [inline crates/runtime/src/embeddings/table.rs]:
```suggestion
                tracing::debug!("Model {model} not found for column {col}");
```

**Comment by Jeadie** [comment]
[Bug 2](https://github.com/spiceai/spiceai/pull/1576#issuecomment-2143814615) is caused by [Bug 1](https://github.com/spiceai/spiceai/pull/1576#issuecomment-2143814087). To verify this. add a `refresh_sql` that has a LIMIT clause, e.g. 
```yaml
datasets:
 - from: s3://spiceai-demo-datasets/cleaned_sales_data.parquet
   name: embeded_cleaned_sales_data
   acceleration:
     enabled: true
     refresh_sql: Select * from embeded_cleaned_sales_data limit 10
   embeddings:
     - column: address_line1
       use: oai
```

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**